rcl_logging_spdlog/src/rcl_logging_spdlog.cpp

```diff
@@ -0,0 +1,154 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cerrno>
+#include <inttypes.h>
+
+#include <rcutils/allocator.h>
+#include <rcutils/filesystem.h>
+#include <rcutils/get_env.h>
+#include <rcutils/process.h>
+#include <rcutils/time.h>
+
+#include "spdlog/spdlog.h"
+#include "spdlog/sinks/basic_file_sink.h"
+
+#include "rcl_logging_spdlog/logging_interface.h"
+
+#define RC_LOGGING_RET_OK                          (0)
+#define RC_LOGGING_RET_ERROR                       (2)
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+static std::shared_ptr<spdlog::logger> g_root_logger;
+
+/* These are defined here to match the severity levels in rcl. They provide a consistent way for external logger
+    implementations to map between the incoming integer severity from ROS to the concept of DEBUG, INFO, WARN, ERROR,
+    and FATAL*/
+enum RC_LOGGING_LOG_SEVERITY
+{
+  RC_LOGGING_SEVERITY_UNSET = 0,  ///< The unset log level
+  RC_LOGGING_SEVERITY_DEBUG = 10,  ///< The debug log level
+  RC_LOGGING_SEVERITY_INFO = 20,  ///< The info log level
+  RC_LOGGING_SEVERITY_WARN = 30,  ///< The warn log level
+  RC_LOGGING_SEVERITY_ERROR = 40,  ///< The error log level
+  RC_LOGGING_SEVERITY_FATAL = 50,  ///< The fatal log level
+};
+
+static spdlog::level::level_enum map_external_log_level_to_library_level(int external_level)
+{
+  spdlog::level::level_enum level;
+
+  // map to the next highest level of severity
+  if (external_level <= RC_LOGGING_SEVERITY_DEBUG) {
+    level = spdlog::level::level_enum::debug;
+  } else if (external_level <= RC_LOGGING_SEVERITY_INFO) {
+    level = spdlog::level::level_enum::info;
+  } else if (external_level <= RC_LOGGING_SEVERITY_WARN) {
+    level = spdlog::level::level_enum::warn;
+  } else if (external_level <= RC_LOGGING_SEVERITY_ERROR) {
+    level = spdlog::level::level_enum::err;
+  } else if (external_level <= RC_LOGGING_SEVERITY_FATAL) {
+    level = spdlog::level::level_enum::critical;
+  }
+  return level;
+}
+
+rcl_logging_ret_t rcl_logging_external_initialize(const char * config_file, rcutils_allocator_t allocator)
+{
+  bool config_file_provided = (nullptr != config_file) && (config_file[0] != '\0');
+  if (config_file_provided) {
+    // TODO(clalancette): deal with the config file
```

Comment:
We could return some sort of NOT_IMPLEMENTED error code here so that it fails if someone tries to provide a config, but I think it'd be just as good to remove this if block for now and just mark the config_file as unused.

Reply:
I'm now using the `RC_LOGGING_RET_ERROR` return here as a stand-in for "this isn't supported yet".  That's on the theory that it is better to fail fast rather than to silently ignore what the user asked for.  So I'm going to mark this as resolved for now, but feel free to make a new comment if you feel we should do something different here.